vscode: fix rpath for native modules

eadwu opened the pull request:
###### Motivation for this change
Affects at least VSCode Insiders 1544768411 and 1545113902.
Current VSCode stable 1.30.1 doesn't need this.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

eadwu committed:
vscode: fix rpath for native modules

eadwu commented on pkgs/applications/editors/vscode/default.nix:
```diff
@@ -22,6 +22,7 @@ let
   rpath = lib.concatStringsSep ":" [
     atomEnv.libPath
     "${lib.makeLibraryPath [gtk2]}"
+    "${lib.makeLibraryPath [at_spi2_atk]}"
```
This seems to be necessary for a correct RPATH patch? If I don't add it there, ldd outputs multiple libatk-bridge-2.0.so.0 entries.

eadwu commented on pkgs/applications/editors/vscode/default.nix:
```diff
@@ -84,14 +85,8 @@ in
-        --set-rpath "${rpath}" \
-        "$out/lib/vscode/resources/app/node_modules.asar.unpacked/native-keymap/build/Release/\
-      keymapping.node"
+      find $out/lib/vscode/resources/app/node_modules.asar.unpacked -type f -name "*.node" \
```
Should I expand this to every module instead of using find? Most of the native node modules need to be patched with `libstdc++.so.6`.

Mic92 replied:
You can have a look at `autoPatchelfHook`. This should patch most files automatically.

veprbl commented on pkgs/applications/editors/vscode/default.nix:
```diff
@@ -1,5 +1,5 @@
 { stdenv, lib, fetchurl, unzip, atomEnv, makeDesktopItem,
-  gtk2, wrapGAppsHook, libXScrnSaver, libxkbfile, libsecret,
+  gtk2, at_spi2_atk, wrapGAppsHook, libXScrnSaver, libxkbfile, libsecret,
```
`at_spi2_atk` is an alias, please use `at-spi2-atk`

Mic92 commented on pkgs/applications/editors/vscode/default.nix:
```diff
@@ -22,6 +22,7 @@ let
+    "${lib.makeLibraryPath [at_spi2_atk]}"
     "${lib.makeLibraryPath [libsecret]}/libsecret-1.so.0"
     "${lib.makeLibraryPath [libXScrnSaver]}/libXss.so.1"
     "${lib.makeLibraryPath [libxkbfile]}/libxkbfile.so.1"
```
I wonder why there are absolute paths to files here. the rpath usually contain directories.

eadwu commented:
Oh, looks like the Insiders build works now, though the stable build is now broken, `ldd` seems to be showing everything got patched and `code --verbose` shows nothing. Gonna look at it again when I have more time.

Mic92 commented on pkgs/applications/editors/vscode/default.nix:
```diff
@@ -53,9 +44,20 @@ in
+        libsecret
+        at-spi2-atk
+        libXScrnSaver
+        autoPatchelfHook
```
autoPatchelfHook is only available on linux. In fact only unzip would be useful on macOS.

Mic92 commented on pkgs/applications/editors/vscode/default.nix:
```diff
@@ -53,9 +44,20 @@ in
-    buildInputs = if stdenv.hostPlatform.system == "x86_64-darwin"
-      then [ unzip libXScrnSaver libsecret ]
-      else [ wrapGAppsHook libXScrnSaver libxkbfile libsecret ];
+    buildInputs = (if stdenv.hostPlatform.system == "x86_64-darwin"
```
Same as `stdenv.isDarwin`.

eadwu commented:
Alright it seems like it might be due to an RPath issue since it seems to use the `libgcc_s.so.1` from `glibc` over `gcc`. The interpreter looks to be fine. Any idea on how to put `gcc` on a higher priority than `glibc` for `autoPatchelfHook`? Tried adding `gcc` as a build input and manually patching it afterwards with `patchelf --set-rpath "${lib.makeLibraryPath[gcc]}:$(patchelf --print-rpath $out/lib/vscode/${executableName})"`.

From old to `autoPatchelfHook`, `libgcc_s.so.1 => /nix/store/x6inizi5ahlyhqxxwv1rvn05a25icarq-gcc-7.3.0-lib/lib/libgcc_s.so.1 (0x00007f70c248b000)` to `libgcc_s.so.1 => /nix/store/xdsjx0gba4id3yyqxv66bxnm2sqixkjj-glibc-2.27/lib/libgcc_s.so.1 (0x00007f5b88498000)`.

Mic92 commented:
@aszlig might know

eadwu commented:
Looks like it's a problem with RPATH since patching in systemd and fontconfig into the RPATH allows it to work. autoPatchelfHook doesn't do it automatically since it doesn't even seem to be required though it does exist when I run ldd. `libsystemd.so.0` only seems to be needed somewhere after the patching.

original ldd output
```
       linux-vdso.so.1 (0x00007ffff7ffa000)
        libnode.so => /nix/store/9hamr71qgjbgr4ld58vvhx2kxzh6a0xs-vscode-1.30.1/lib/vscode/libnode.so (0x00007ffff6b03000)
        libpthread.so.0 => /nix/store/xdsjx0gba4id3yyqxv66bxnm2sqixkjj-glibc-2.27/lib/libpthread.so.0 (0x00007ffff68e4000)
        libgtk-3.so.0 => not found
        libgdk-3.so.0 => not found
        libpangocairo-1.0.so.0 => not found
        libpango-1.0.so.0 => not found
        libatk-1.0.so.0 => not found
        libcairo.so.2 => not found
        libgdk_pixbuf-2.0.so.0 => not found
        libgio-2.0.so.0 => not found
        libgobject-2.0.so.0 => not found
        libdbus-1.so.3 => not found
        libX11-xcb.so.1 => not found
        libxcb.so.1 => not found
        libXi.so.6 => not found
        libXcursor.so.1 => not found
        libXdamage.so.1 => not found
        libXrandr.so.2 => not found
        libXcomposite.so.1 => not found
        libXext.so.6 => not found
        libXfixes.so.3 => not found
        libXrender.so.1 => not found
        libX11.so.6 => not found
        libXtst.so.6 => not found
        libXss.so.1 => not found
        libgconf-2.so.4 => not found
        libgmodule-2.0.so.0 => not found
        libglib-2.0.so.0 => not found
        libnss3.so => not found
        libnssutil3.so => not found
        libsmime3.so => not found
        libnspr4.so => not found
        libffmpeg.so => /nix/store/9hamr71qgjbgr4ld58vvhx2kxzh6a0xs-vscode-1.30.1/lib/vscode/libffmpeg.so (0x00007ffff7e08000)
        libasound.so.2 => not found
        libcups.so.2 => not found
        librt.so.1 => /nix/store/xdsjx0gba4id3yyqxv66bxnm2sqixkjj-glibc-2.27/lib/librt.so.1 (0x00007ffff66dc000)
        libdl.so.2 => /nix/store/xdsjx0gba4id3yyqxv66bxnm2sqixkjj-glibc-2.27/lib/libdl.so.2 (0x00007ffff64d8000)
        libfontconfig.so.1 => not found
        libexpat.so.1 => not found
        libstdc++.so.6 => not found
        libm.so.6 => /nix/store/xdsjx0gba4id3yyqxv66bxnm2sqixkjj-glibc-2.27/lib/libm.so.6 (0x00007ffff6143000)
        libgcc_s.so.1 => /nix/store/xdsjx0gba4id3yyqxv66bxnm2sqixkjj-glibc-2.27/lib/libgcc_s.so.1 (0x00007ffff5f2d000)
        libc.so.6 => /nix/store/xdsjx0gba4id3yyqxv66bxnm2sqixkjj-glibc-2.27/lib/libc.so.6 (0x00007ffff5b79000)
        /lib64/ld-linux-x86-64.so.2 => /nix/store/xdsjx0gba4id3yyqxv66bxnm2sqixkjj-glibc-2.27/lib64/ld-linux-x86-64.so.2 (0x00007ffff7dd8000)
        libstdc++.so.6 => not found
```

after build
```
        linux-vdso.so.1 (0x00007ffcc3377000)
        libnode.so => /nix/store/9hamr71qgjbgr4ld58vvhx2kxzh6a0xs-vscode-1.30.1/lib/vscode/libnode.so (0x00007f36370b7000)
        libpthread.so.0 => /nix/store/xdsjx0gba4id3yyqxv66bxnm2sqixkjj-glibc-2.27/lib/libpthread.so.0 (0x00007f3636e98000)
        libgtk-3.so.0 => /nix/store/jjkzzq4ngyar229mdkis3yswcsil9w38-gtk+3-3.24.1/lib/libgtk-3.so.0 (0x00007f3636559000)
        libgdk-3.so.0 => /nix/store/jjkzzq4ngyar229mdkis3yswcsil9w38-gtk+3-3.24.1/lib/libgdk-3.so.0 (0x00007f363626b000)
        libpangocairo-1.0.so.0 => /nix/store/jlrqghvz9wabqnrcrpc0v3jfh7yz3n2w-pango-1.42.4/lib/libpangocairo-1.0.so.0 (0x00007f363605d000)
        libpango-1.0.so.0 => /nix/store/jlrqghvz9wabqnrcrpc0v3jfh7yz3n2w-pango-1.42.4/lib/libpango-1.0.so.0 (0x00007f3635e17000)
        libatk-1.0.so.0 => /nix/store/yiq6ba061hgx4bqq7d34zvgn5rj6xs5s-atk-2.30.0/lib/libatk-1.0.so.0 (0x00007f3635bef000)
        libcairo.so.2 => /nix/store/jp5pzj831wh3dwhyalm9mzkyvhqzxhb4-cairo-1.16.0/lib/libcairo.so.2 (0x00007f36358b8000)
        libgdk_pixbuf-2.0.so.0 => /nix/store/8lar1j9z7i9jd4d3fqp3pfi7x1xdmaxn-gdk-pixbuf-2.38.0/lib/libgdk_pixbuf-2.0.so.0 (0x00007f363568a000)
        libgio-2.0.so.0 => /nix/store/1m73xy2036f3lpyizgmsclab9si27lxj-glib-2.58.1/lib/libgio-2.0.so.0 (0x00007f36352bd000)
        libgobject-2.0.so.0 => /nix/store/1m73xy2036f3lpyizgmsclab9si27lxj-glib-2.58.1/lib/libgobject-2.0.so.0 (0x00007f3635064000)
        libdbus-1.so.3 => /nix/store/7abmm73058b725dkb6asmqnxidknsd99-dbus-1.12.12-lib/lib/libdbus-1.so.3 (0x00007f3634e12000)
        libX11-xcb.so.1 => /nix/store/xfd6ip748l4vkym372s7sww1x22nwr06-libX11-1.6.7/lib/libX11-xcb.so.1 (0x00007f3634c10000)
        libxcb.so.1 => /nix/store/ji0fkcgpqwvb65l14wx9r1n8zm6n7d9s-libxcb-1.13.1/lib/libxcb.so.1 (0x00007f36349e7000)
        libXi.so.6 => /nix/store/yxa777qj7b0lrxpa2qx20zr74h2vpj0g-libXi-1.7.9/lib/libXi.so.6 (0x00007f36347d6000)
        libXcursor.so.1 => /nix/store/nfpy7bg74n84ihpg1a56qnx734y09ap7-libXcursor-1.1.15/lib/libXcursor.so.1 (0x00007f36345cb000)
        libXdamage.so.1 => /nix/store/5bhm4dclsmizxgpwrzjzsv8c1prqqlh4-libXdamage-1.1.4/lib/libXdamage.so.1 (0x00007f36343c8000)
        libXrandr.so.2 => /nix/store/dv5i0p8844fl6a9rzw1bcmckga9wp2ny-libXrandr-1.5.1/lib/libXrandr.so.2 (0x00007f36341bd000)
        libXcomposite.so.1 => /nix/store/vmvljj56nqf5yil3nxjymvrjynz0cd5z-libXcomposite-0.4.4/lib/libXcomposite.so.1 (0x00007f3633fba000)
        libXext.so.6 => /nix/store/0ympzp3qwdkq1ms8nbv5fv06yj5ayjdl-libXext-1.3.3/lib/libXext.so.6 (0x00007f3633da7000)
        libXfixes.so.3 => /nix/store/jcycns512pa5g44m5wnnca0g7ssg76ai-libXfixes-5.0.3/lib/libXfixes.so.3 (0x00007f3633ba1000)
        libXrender.so.1 => /nix/store/ksjmabwc3nwhmv1h1xgk9vi2q0gq63x5-libXrender-0.9.10/lib/libXrender.so.1 (0x00007f3633996000)
        libX11.so.6 => /nix/store/xfd6ip748l4vkym372s7sww1x22nwr06-libX11-1.6.7/lib/libX11.so.6 (0x00007f3633655000)
        libXtst.so.6 => /nix/store/7fsv1zayxhcbgk72m99f2672m5w2q1q6-libXtst-1.2.3/lib/libXtst.so.6 (0x00007f363344e000)
        libXss.so.1 => /nix/store/97bn75j0w1f5hn7893y4911c61lrmkmf-libXScrnSaver-1.2.2/lib/libXss.so.1 (0x00007f363324a000)
        libgconf-2.so.4 => /nix/store/3y7dbj0v1ykiyi9wjb11xsncafwr98qv-gconf-3.2.6/lib/libgconf-2.so.4 (0x00007f3633009000)
        libgmodule-2.0.so.0 => /nix/store/1m73xy2036f3lpyizgmsclab9si27lxj-glib-2.58.1/lib/libgmodule-2.0.so.0 (0x00007f3632e05000)
        libglib-2.0.so.0 => /nix/store/1m73xy2036f3lpyizgmsclab9si27lxj-glib-2.58.1/lib/libglib-2.0.so.0 (0x00007f3632ac8000)
        libnss3.so => /nix/store/za3jl9j69f6bplz335hy10bsmypqd8vx-nss-3.40.1/lib/libnss3.so (0x00007f363279c000)
        libnssutil3.so => /nix/store/za3jl9j69f6bplz335hy10bsmypqd8vx-nss-3.40.1/lib/libnssutil3.so (0x00007f363256b000)
        libsmime3.so => /nix/store/za3jl9j69f6bplz335hy10bsmypqd8vx-nss-3.40.1/lib/libsmime3.so (0x00007f3632342000)
        libnspr4.so => /nix/store/x7x60wfm20rymhsazqlim1xc53mnn0fk-nspr-4.20/lib/libnspr4.so (0x00007f3632102000)
        libffmpeg.so => /nix/store/9hamr71qgjbgr4ld58vvhx2kxzh6a0xs-vscode-1.30.1/lib/vscode/libffmpeg.so (0x00007f36384b3000)
        libasound.so.2 => /nix/store/yxvgvxrbhl6xrj65h7gy73ab2k9g60r9-alsa-lib-1.1.7/lib/libasound.so.2 (0x00007f3631e00000)
        libcups.so.2 => /nix/store/kbfrnb2mxkm5y2lnbvdfjdjinjw8b4n3-cups-2.2.6-lib/lib/libcups.so.2 (0x00007f3631b82000)
        librt.so.1 => /nix/store/xdsjx0gba4id3yyqxv66bxnm2sqixkjj-glibc-2.27/lib/librt.so.1 (0x00007f363197a000)
        libdl.so.2 => /nix/store/xdsjx0gba4id3yyqxv66bxnm2sqixkjj-glibc-2.27/lib/libdl.so.2 (0x00007f3631776000)
        libfontconfig.so.1 => /nix/store/ngipd3p6yd5l82af22hs6i3n5mswk1pa-fontconfig-2.12.6-lib/lib/libfontconfig.so.1 (0x00007f3631530000)
        libexpat.so.1 => /nix/store/dxiswg6gy3m4x32syhwwg7w0mj6i6jig-expat-2.2.6/lib/libexpat.so.1 (0x00007f36312fe000)
        libstdc++.so.6 => /nix/store/x6inizi5ahlyhqxxwv1rvn05a25icarq-gcc-7.3.0-lib/lib/libstdc++.so.6 (0x00007f3630f77000)
        libm.so.6 => /nix/store/xdsjx0gba4id3yyqxv66bxnm2sqixkjj-glibc-2.27/lib/libm.so.6 (0x00007f3630be2000)
        libgcc_s.so.1 => /nix/store/x6inizi5ahlyhqxxwv1rvn05a25icarq-gcc-7.3.0-lib/lib/libgcc_s.so.1 (0x00007f36309ca000)
        libc.so.6 => /nix/store/xdsjx0gba4id3yyqxv66bxnm2sqixkjj-glibc-2.27/lib/libc.so.6 (0x00007f3630616000)
        /nix/store/xdsjx0gba4id3yyqxv66bxnm2sqixkjj-glibc-2.27/lib/ld-linux-x86-64.so.2 => /nix/store/xdsjx0gba4id3yyqxv66bxnm2sqixkjj-glibc-2.27/lib64/ld-linux-x86-64.so.2 (0x00007f3638480000)
        libXinerama.so.1 => /nix/store/1q1g74swhk2h60b5bidnxnrdd4hphspb-libXinerama-1.1.4/lib/libXinerama.so.1 (0x00007f3630413000)
        libcairo-gobject.so.2 => /nix/store/jp5pzj831wh3dwhyalm9mzkyvhqzxhb4-cairo-1.16.0/lib/libcairo-gobject.so.2 (0x00007f363020a000)
        libpixman-1.so.0 => /nix/store/snvbfmvzy0cq8r1zamnsi34nwrfkm839-pixman-0.36.0/lib/libpixman-1.so.0 (0x00007f362ff62000)
        libEGL.so.1 => /nix/store/azvjiz14if7dinh00ai50d7lq93qwx34-libglvnd-1.0.0/lib/libEGL.so.1 (0x00007f362fd4e000)
        libxcb-shm.so.0 => /nix/store/ji0fkcgpqwvb65l14wx9r1n8zm6n7d9s-libxcb-1.13.1/lib/libxcb-shm.so.0 (0x00007f362fb4a000)
        libxcb-render.so.0 => /nix/store/ji0fkcgpqwvb65l14wx9r1n8zm6n7d9s-libxcb-1.13.1/lib/libxcb-render.so.0 (0x00007f362f93c000)
        libGL.so.1 => /nix/store/azvjiz14if7dinh00ai50d7lq93qwx34-libglvnd-1.0.0/lib/libGL.so.1 (0x00007f362f6ab000)
        libGLX.so.0 => /nix/store/azvjiz14if7dinh00ai50d7lq93qwx34-libglvnd-1.0.0/lib/libGLX.so.0 (0x00007f362f479000)
        libXau.so.6 => /nix/store/5jnnzqfxf6y7kq6npjqnrgn3c1j61xzm-libXau-1.0.8/lib/libXau.so.6 (0x00007f362f275000)
        libXdmcp.so.6 => /nix/store/03qq9wvzbfkjail9ycwmn0f3sb34182y-libXdmcp-1.1.2/lib/libXdmcp.so.6 (0x00007f362f06f000)
        libGLdispatch.so.0 => /nix/store/azvjiz14if7dinh00ai50d7lq93qwx34-libglvnd-1.0.0/lib/libGLdispatch.so.0 (0x00007f362edb9000)
        libatk-bridge-2.0.so.0 => /nix/store/nbsz41kfcd3k9jxf370lqbl42c23yz8a-at-spi2-atk-2.30.0/lib/libatk-bridge-2.0.so.0 (0x00007f362eb82000)
        libxkbcommon.so.0 => /nix/store/1gapf57gss75cqsdlk5imzjkjzd7g4dz-libxkbcommon-0.8.0/lib/libxkbcommon.so.0 (0x00007f362e932000)
        libwayland-cursor.so.0 => /nix/store/9a075za4462rbn75hkzsdlac77af7w5z-wayland-1.16.0/lib/libwayland-cursor.so.0 (0x00007f362e72a000)
        libwayland-egl.so.1 => /nix/store/9a075za4462rbn75hkzsdlac77af7w5z-wayland-1.16.0/lib/libwayland-egl.so.1 (0x00007f362e528000)
        libwayland-client.so.0 => /nix/store/9a075za4462rbn75hkzsdlac77af7w5z-wayland-1.16.0/lib/libwayland-client.so.0 (0x00007f362e318000)
        libffi.so.6 => /nix/store/ddgaziqy9m2jgm4z8dzkhqd3w9hzryns-libffi-3.2.1/lib/../lib64/libffi.so.6 (0x00007f362e10e000)
        libepoxy.so.0 => /nix/store/fnabfd42az2v4pwhv754fi4v0bjiqzb0-epoxy-1.5.2/lib/libepoxy.so.0 (0x00007f362ddd9000)
        libpangoft2-1.0.so.0 => /nix/store/jlrqghvz9wabqnrcrpc0v3jfh7yz3n2w-pango-1.42.4/lib/libpangoft2-1.0.so.0 (0x00007f362dbc3000)
        libharfbuzz.so.0 => /nix/store/zr4k7b7xkw70v2g1vjbaz24wqj9jn2gx-harfbuzz-2.1.0/lib/libharfbuzz.so.0 (0x00007f362d8e0000)
        libgraphite2.so.3 => /nix/store/fv660vamzmzgzab8laq15kdk1kda4jyj-graphite2-1.3.6/lib/libgraphite2.so.3 (0x00007f362d6ad000)
        libgthread-2.0.so.0 => /nix/store/1m73xy2036f3lpyizgmsclab9si27lxj-glib-2.58.1/lib/libgthread-2.0.so.0 (0x00007f362d4ab000)
        libfribidi.so.0 => /nix/store/5442hl9j0s9lnapv1rp89xr9fdszfgiy-fribidi-1.0.5/lib/libfribidi.so.0 (0x00007f362d28e000)
        libfreetype.so.6 => /nix/store/b0pqjv07ky71jlw9bjnw21lgyq39057q-freetype-2.9/lib/libfreetype.so.6 (0x00007f362cfd5000)
        libbz2.so.1 => /nix/store/40ql7v4ai08wm61xps1kh5s4issc3yb8-bzip2-1.0.6.0.1/lib/libbz2.so.1 (0x00007f362cdc5000)
        libpng16.so.16 => /nix/store/px5w0kjsn31dcs1zscqss8zjxkb9j45m-libpng-apng-1.6.35/lib/libpng16.so.16 (0x00007f362cb8e000)
        libz.so.1 => /nix/store/nsrwlly5ikwhppcslaazp547kh7q7j6s-zlib-1.2.11/lib/libz.so.1 (0x00007f362c972000)
        libmount.so.1 => /nix/store/z0x40kmai1daywwv9v1p2d1wvc6wzhpw-util-linux-2.33/lib/libmount.so.1 (0x00007f362c714000)
        libselinux.so.1 => /nix/store/v4i35a0bhbxb146vsbd8fd66wg74zqk4-libselinux-2.7/lib/libselinux.so.1 (0x00007f362c4e9000)
        libresolv.so.2 => /nix/store/xdsjx0gba4id3yyqxv66bxnm2sqixkjj-glibc-2.27/lib/libresolv.so.2 (0x00007f362c2d2000)
        libsystemd.so.0 => /nix/store/wf8gwznvvsbyb54m6qxmvh02c8gv94dy-systemd-239-lib/lib/libsystemd.so.0 (0x00007f362c021000)
        libdbus-glib-1.so.2 => /nix/store/bgbdk8pbirvvql46m2bgm64fkawvfniy-dbus-glib-0.110/lib/libdbus-glib-1.so.2 (0x00007f362bdf8000)
        libORBit-2.so.0 => /nix/store/880pmbbkfpwzlbzhjz5gngbpmqbgqxj2-ORBit2-2.14.19/lib/libORBit-2.so.0 (0x00007f362bb86000)
        libpcre.so.1 => /nix/store/5l37h12ga8paxryhd0sf7j2wbcj92f79-pcre-8.42/lib/libpcre.so.1 (0x00007f362b912000)
        libplc4.so => /nix/store/x7x60wfm20rymhsazqlim1xc53mnn0fk-nspr-4.20/lib/libplc4.so (0x00007f362b70d000)
        libplds4.so => /nix/store/x7x60wfm20rymhsazqlim1xc53mnn0fk-nspr-4.20/lib/libplds4.so (0x00007f362b509000)
        libgnutls.so.30 => /nix/store/6s135y1341ip077r6r8p0ifkb0kqfck2-gnutls-3.6.2/lib/libgnutls.so.30 (0x00007f362b1ab000)
        libavahi-common.so.3 => /nix/store/w60vmi7nfd291dp0pbrxcz8xb73bsajy-avahi-0.7/lib/libavahi-common.so.3 (0x00007f362af9c000)
        libavahi-client.so.3 => /nix/store/w60vmi7nfd291dp0pbrxcz8xb73bsajy-avahi-0.7/lib/libavahi-client.so.3 (0x00007f362ad8a000)
        libcrypt.so.1 => /nix/store/xdsjx0gba4id3yyqxv66bxnm2sqixkjj-glibc-2.27/lib/libcrypt.so.1 (0x00007f362ab52000)
        libatspi.so.0 => /nix/store/72qi1zp1zyw809ppdmpynn8pgcgh7im8-at-spi2-core-2.30.0/lib/libatspi.so.0 (0x00007f362a917000)
        libblkid.so.1 => /nix/store/z0x40kmai1daywwv9v1p2d1wvc6wzhpw-util-linux-2.33/lib/libblkid.so.1 (0x00007f362a6c3000)
        libuuid.so.1 => /nix/store/z0x40kmai1daywwv9v1p2d1wvc6wzhpw-util-linux-2.33/lib/libuuid.so.1 (0x00007f362a4bb000)
        liblzma.so.5 => /nix/store/430ajcrv4llfcw12kgizgkb5rnm2hczk-xz-5.2.4/lib/liblzma.so.5 (0x00007f362a294000)
        liblz4.so.1 => /nix/store/a9gwz05611m6rxdc55hpla5zns3h36xw-lz4-1.8.3/lib/liblz4.so.1 (0x00007f362a05f000)
        libcap.so.2 => /nix/store/r5g3f78k3qcrqj05bl1wv04c12r589bl-libcap-2.25-lib/lib/libcap.so.2 (0x00007f3629e5a000)
        libgcrypt.so.20 => /nix/store/jycv1dmyjs3045rsnbky3m1s7xdj7xpg-libgcrypt-1.8.4/lib/libgcrypt.so.20 (0x00007f3629b3d000)
        libp11-kit.so.0 => /nix/store/yr1fdkyz2gxnfwifn1q2pifgl0k54yab-p11-kit-0.23.14/lib/libp11-kit.so.0 (0x00007f362980f000)
        libunistring.so.2 => /nix/store/ja2h63lw0x1mvzyam4hqv2iwvvcwa8gm-libunistring-0.9.10/lib/libunistring.so.2 (0x00007f362948a000)
        libtasn1.so.6 => /nix/store/0qjvyh5q1yd16knxa9xh15pnrx4lb61w-libtasn1-4.13/lib/libtasn1.so.6 (0x00007f3629277000)
        libnettle.so.6 => /nix/store/yskk4g4x20d6gk867hdcwzn81lkqkwwr-nettle-3.4/lib/libnettle.so.6 (0x00007f362903f000)
        libhogweed.so.4 => /nix/store/yskk4g4x20d6gk867hdcwzn81lkqkwwr-nettle-3.4/lib/libhogweed.so.4 (0x00007f3628e08000)
        libgmp.so.10 => /nix/store/9yb6gizxpll4l6gf7md75kbii6b1dvjc-gmp-6.1.2/lib/libgmp.so.10 (0x00007f3628b74000)
        libssp.so.0 => /nix/store/x6inizi5ahlyhqxxwv1rvn05a25icarq-gcc-7.3.0-lib/lib/libssp.so.0 (0x00007f362896f000)
        libgpg-error.so.0 => /nix/store/kvd0aw4ikb85an2fjsx7ldp20ji11p1z-libgpg-error-1.32/lib/libgpg-error.so.0 (0x00007f362874f000)
```

aszlig commented:
> The interpreter looks to be fine. Any idea on how to put gcc on a higher priority than glibc for autoPatchelfHook?

There is no direct way to assign priorities, but you can give a dependency precedence over others by using `addAutoPatchelfSearchPath "$gcclibs"` before `postFixup`.

Mic92 commented on pkgs/applications/editors/vscode/default.nix:
```diff
@@ -79,19 +73,8 @@ in
-        "$out/lib/vscode/resources/app/node_modules.asar.unpacked/native-keymap/build/Release/\
-      keymapping.node"
+      wrapProgram $out/lib/vscode/${executableName} \
+        --prefix LD_LIBRARY_PATH : ${atomEnv.libPath}
```
Are those libraries loaded via `dlopen`?

Mic92 replied:
If so you can add those manually: https://github.com/NixOS/nixpkgs/pull/50596/files#diff-cdcfceb5de8d7eaedeec7ca699f6b918R15
This is still better then adding all libraries to LD_LIBRARY_PATH because some other programs that are started by vscode might break if incompatible libraries are preloaded.

eadwu replied:
Actually I think `$out/lib/vscode/${executableName}` is just a renaming of the `electron` executable so it might be better to link the one generated from `electron`?

worldofpeace commented:
https://github.com/NixOS/nixpkgs/pull/55362 needs this.

@Mic92 Any other reservations about this?

worldofpeace committed:
vscode: minor cleanup, produce one wrapper on linux

I removed the hack for libsecret, introduced in[0] and
I didn't encounter any problems at runtime.

[0]: https://github.com/NixOS/nixpkgs/pull/29127

worldofpeace commented:
@eadwu I can merge this if you can verify https://github.com/NixOS/nixpkgs/pull/52419/commits/351306304935d1d54dca0a88e2fe9ab590da75dc doesn't break anything for you.

eadwu commented:
Works fine for me, also a crude test of `find -type f -executable -exec ldd "{}" \;` shows all dependencies are found.
